### PR TITLE
Editable FP Registers

### PR DESCRIPTION
**Pull request by cayb0rg**
Merged Visual Datapath Upgrade #349 into this one because the editable registers relies on the changes to the coprocessor

* Visual datapath now relies on `datapath_state`
* Adds a universal `visual_line_to_data` function to utils
* Adds editable FP registers
* Adds new FPRegisters struct similar to GPRegisters
  * This abstracts away a few of the tasks required by the UI (such as checking valid register values)
* Compresses reg view functions into one for gpr and one for fpr

### CODE REVIEW

File: src/emulation_core/mips/line_info.rs [hunk 1]
```diff
@@ -330,3 +333,317 @@ impl VisualDatapath for MipsDatapath {
         }
     }
 }
+
+pub fn visual_line_to_data(
```
Comment (brooksmckinley) [inline src/emulation_core/mips/line_info.rs:337]: This seems identical to the implementation in visual_datapath/utils.rs, can one of them be removed?

File: src/emulation_core/mips/datapath.rs [hunk 2]
```diff
@@ -277,6 +277,11 @@ impl Datapath for MipsDatapath {
         *register = _data;
     }
 
+    fn set_fp_register_by_str(&mut self, _register: &str, _data: Self::RegisterData) {
+        let register = &mut self.coprocessor.registers[_register];
```
Comment (brooksmckinley) [inline src/emulation_core/mips/datapath.rs:281]: This one's probably my fault from when I wrote the datapath trait, but can we remove the underscores in the argument names? I used them back in the datapath trait when I was writing to turn off the unused variable warnings, but they're not really needed anymore. There's a few other places in this PR that have underscored arguments that end up getting used too.

**Review by brooksmckinley — CHANGES_REQUESTED** [review]
Amazing PR! All I really have are small simplifications and nitpicks.

**Comment by cayb0rg** [comment]
Thank you for the feedback!

**Review by brooksmckinley — APPROVED** [review]
LGTM, ready to merge if you are.

**Pull request merged**